src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1617 @@
+                    << num_frames << " frames in this arc. ";
+
+      auto new_prefinal_state = lat_.AddState();
+      forward_costs_.resize(new_prefinal_state + 1);
+      forward_costs_[new_prefinal_state] = forward_costs_[prefinal_state];
+
+      std::vector<CompactLatticeArc> arcs_remained;
+      for (aiter.Reset(); !aiter.Done(); aiter.Next()) {
+        auto arc = aiter.Value();
+        bool remain_the_arc = true; // If we remain the arc, the state will not be
+                                    // re-determinized, vice versa.
+        if (arc.olabel > config_.max_word_id) { // final arc
+          KALDI_ASSERT(arc.olabel < state_last_initial_offset_);
+          KALDI_ASSERT(lat_.Final(arc.nextstate) != CompactLatticeWeight::Zero());
+          remain_the_arc = false;
+        } else {
+          int num_frames_exclude_arc = num_frames - arc.weight.String().size();
+          // destination-state of the arc is further than redeterminize_max_frames
+          // from the most recent frame we are determinizing
+          if (num_frames_exclude_arc > config_.redeterminize_max_frames)
+            remain_the_arc = true;
+          else {
+            // destination-state of the arc is no further than
+            // redeterminize_max_frames from the most recent frame we are
+            // determinizing
+            auto r = final_arc_list_prev_.find(arc.nextstate);
+            // destination-state of the arc is not prefinal state
+            if (r == final_arc_list_prev_.end()) remain_the_arc = true;
+            // destination-state of the arc is prefinal state
+            else
+              remain_the_arc = false;
+          }
+        }
+
+        if (remain_the_arc)
+          arcs_remained.push_back(arc);
+        else
+          lat_.AddArc(new_prefinal_state, arc);
+      }
+      CompactLatticeArc arc_to_new(0, 0, CompactLatticeWeight::One(),
+                                   new_prefinal_state);
+      arcs_remained.push_back(arc_to_new);
+
+      lat_.DeleteArcs(prefinal_state);
+      for (auto &i : arcs_remained) lat_.AddArc(prefinal_state, i);
+      processed_prefinal_states_[prefinal_state] = new_prefinal_state;
+    }
+  }
+  KALDI_VLOG(8) << "states of the lattice after GetRedeterminizedStates: "
+                << lat_.NumStates();
+}
+
+// This function is specifically designed to obtain the initial arcs for a chunk
+// We have multiple states for one token label after determinization
+template <typename FST>
+void LatticeIncrementalDeterminizer<FST>::GetInitialRawLattice(
+    Lattice *olat,
+    unordered_multimap<int, LatticeArc::StateId> *token_label2last_state_map,
+    const unordered_map<int32, BaseFloat> &token_label_final_cost) {
+  using namespace fst;
+  typedef LatticeArc Arc;
+  typedef Arc::StateId StateId;
+  typedef Arc::Weight Weight;
+  typedef Arc::Label Label;
+
+  GetRedeterminizedStates();
+
+  olat->DeleteStates();
```

Comment:
It seems to me that when this function is called, olat should already be cleared, at least in the use case provided in this code. 

But maybe this is OK -- I guess this function could be called by the user.